### PR TITLE
Move LoggerProvider and friends (OTEL1000) to a stable API

### LINKED ISSUE #4014: Improve DI friendliness of Logging
# Feature Request

**Is your feature request related to a problem?**

- Today Logging extensions provide `AddOpenTelemetry` extension on `ILoggingBuilder`, however `AddProcessor` and `SetResourceProvider` extensions are exposed on `OpenTelemetryLoggerOptions`. This is restrictive for services which are primarily using DI. Exposing these extensions on `OpenTelemetryLoggerOptions` and not on `ILoggingBuilder` and/or `IServiceCollection` limits the ability for third party libraries to provide easier capabilities for adding processor for example, or to compose the resource provider from resources exposed by multiple third party libraries etc. With the DI based mechanism a processor can then be added which can get access to other objects via DI injection and making the APIs really DI friendly for users
- Another ask here is to also expose an override for the existing `AddOpenTelemetry` extension to take in options via 

**Describe the solution you'd like:**

Expose the extensions `SetResourceProvider` and `AddProcessor` on `ILoggingBuilder`.
```csharp
public static ILoggingBuilder AddProcessor<T>(this ILoggingBuilder builder)
            where T : BaseProcessor<LogRecord>{}

public static ILoggingBuilder SetResourceBuilder(this ILoggingBuilder loggingBuilder, ResourceBuilder resourceBuilder);
```

Add another overload for `AddOpenTelemetry` extension
```csharp
public static ILoggingBuilder AddOpenTelemetry(this ILoggingBuilder builder, IConfigurationSection configurationSection) {}
```

**Describe alternatives you've considered.**

Currently we had to implement custom LoggerProvider to get the right functionality

## Additional Context

We would like to move to directly using OpenTelemetryLogger and extending it as opposed to having our own implementation of LoggerProvider and Logger.

### LINKED ISSUE #5442: Promote LoggerProvider from experimental to stable
`LoggerProvider` & `LoggerProviderBuilder` were added as part of the effort to implement the [Log Bridge API](https://github.com/open-telemetry/opentelemetry-dotnet/issues/4433).

Later things were split into two different diagnostics:

* [LoggerProvider - OTEL1000](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/diagnostics/experimental-apis/OTEL1000.md)
* [Log Bridge API - OTEL1001](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/diagnostics/experimental-apis/OTEL1001.md)

This issue is to propose promoting `LoggerProvider` (OTEL1000) to a public API for 1.9.0.

Why do that?

In short, #4014. We don't have the same DI surface in logging that we do for metrics & tracing. `LoggerProvider` & `LoggerProviderBuilder` give us parity in logging with what we have for metrics & tracing.

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #5442
Fixes #4014

## Changes

* Expose the experimental APIs covered by OTEL1000 diagnostic
  * `ILoggingBuilder.UseOpenTelemetry` extension left experimental for now. Pending #5325 it may or may not be exposed at all.
  * `Sdk.CreateLoggerProviderBuilder` moved to `OTEL1001` (remains experimental).

## Migration guidance for library authors

Library authors should begin adding extension methods targeting `LoggerProviderBuilder`. In a future release the `ILoggingBuilder.AddOpenTelemetry` extension method and the `OpenTelemetryLoggerOptions.AddProcessor` & `OpenTelemetryLoggerOptions.SetResourceBuilder` methods will be marked obsolete.

Before:

```csharp
    public static OpenTelemetryLoggerOptions AddMyExporter(
        this OpenTelemetryLoggerOptions loggerOptions,
        Action<MyExporterOptions> configure) {}
```

After:

```csharp
    public static OpenTelemetryLoggerOptions AddMyExporter(
        this OpenTelemetryLoggerOptions loggerOptions,
        Action<MyExporterOptions> configure) {}

    public static LoggerProviderBuilder AddMyExporter(
        this LoggerProviderBuilder builder,
        Action<MyExporterOptions> configure) {}
```

`LoggerProviderBuilder` has the same `AddProcessor` and `SetResourceBuilder` methods previously available on `OpenTelemetryLoggerOptions` as well as the full set of methods\features available to `MeterProviderBuilder` and `TracerProviderBuilder`.

## Migration guidance for users

Note: No migration is currently required for users. All existing code should function as it did before. The following migration paths are purely optional.

### When using OpenTelemetry.Extensions.Hosting

Before:

```csharp
appBuilder.Logging.AddOpenTelemetry(options =>
{
    options.IncludeFormattedMessage = true;
    options.AddOtlpExporter(o => o.Endpoint = new Uri("http://myotlpendpoint:4317/logs"));
});
```

After:

```csharp
appBuilder.Services.AddOpenTelemetry()
    .WithLogging(
        logging => logging.AddOtlpExporter(o => o.Endpoint = new Uri("http://myotlpendpoint:4317/logs")),
        options => options.IncludeFormattedMessage = true);
```

For this to work, your exporter needs to expose an extension which targets `LoggerProviderBuilder`. Extensions are provided for `OtlpExporter`, `ConsoleExporter`, and `InMemoryExporter`.

If you are using an exporter from the https://github.com/open-telemetry/opentelemetry-dotnet-contrib repository or some other exporter, it may not provide this extension yet. First try to upgrade your exporter package. If there isn't an update available, you may use this pattern and continue to call the existing extension (while waiting for the component to be updated):

```csharp
appBuilder.Services.AddOpenTelemetry()
    .WithLogging(
	logging => {},
	options =>
    	{
        	options.IncludeFormattedMessage = true;

        	// Note: Existing extensions targeting OpenTelemetryLoggerOptions can still be called.
        	options.AddGenevaLogExporter(o => o.ConnectionString = "...");
    	});
```

### When using LoggerFactory.Create

There is no dedicated API provided yet to migrate from `ILoggerBuilder.AddOpenTelemetry` (which provides `OpenTelemetryLoggerOptions`) to something which provides `LoggerProviderBuilder`. This API will come in a future version.

If for some reason you want to use `LoggerProviderBuilder` before that API is available, it is possible to use `IServiceCollection.ConfigureOpenTelemetryLoggerProvider` to configure the `LoggerProviderBuilder` using `ILoggingBuilder.Services`:

Before

```csharp
using var loggerFactory = LoggerFactory.Create(builder => builder
    .AddOpenTelemetry(options =>
    {
        options.IncludeFormattedMessage = true;
        options.AddOtlpExporter(o => o.Endpoint = new Uri("http://myotlpendpoint:4317/logs"));
    });
```

After

```csharp
using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.AddOpenTelemetry(options => options.IncludeFormattedMessage = true);

    builder.Services.ConfigureOpenTelemetryLoggerProvider(
        logging => logging.AddConsoleExporter());
});
```

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by CodeBlanch** [comment]
Some thoughts for anyone taking a look at this...

* We don't necessarily have to obsolete _anything_. I put them in initially to see the fallout. I added migration guidance to the PR so we can see what it would look like.

* There is no `Sdk.CreateLoggerProviderBuilder` being added here. I moved that to the Log Bridge experimental API (OTEL1001). The only use case I can think of for `Sdk.CreateLoggerProviderBuilder` is in order to build a `LoggerProvider` manually to feed into a bridge. For the use cases we have today, `LoggerFactory.Create` works fine so I didn't feel this API was needed at this time.

**Comment by vishweshbankwar** [comment]
> Some thoughts for anyone taking a look at this...
> 
> * We don't necessarily have to obsolete _anything_. I put them in initially to see the fallout. I added migration guidance to the PR so we can see what it would look like.
> * There is no `Sdk.CreateLoggerProviderBuilder` being added here. I moved that to the Log Bridge experimental API (OTEL1001). The only use case I can think of for `Sdk.CreateLoggerProviderBuilder` is in order to build a `LoggerProvider` manually to feed into a bridge. For the use cases we have today, `LoggerFactory.Create` works fine so I didn't feel this API was needed at this time.

@CodeBlanch - I think we should not mark things obsolete rn.

Regarding the migration guidance:

1. Migration guidance for case 1:

    ```csharp
    appBuilder.Logging.AddOpenTelemetry(options =>
    {
        options.IncludeFormattedMessage = true;
        options.AddOtlpExporter(o => o.Endpoint = new Uri("http://myotlpendpoint:4317/logs"));
    });
    ```
    
    With this change, my understanding was, they should be able to move to following pattern (bringing consistency with other two signals for hosting scenarios and the update you have in the example)
    
    ```csharp
    appBuilder.Services.AddOpenTelemetry().WithLogging()
    ```

2. Migration guidance for case 2:

    ```csharp
    using var loggerFactory = LoggerFactory.Create(builder => builder
            .UseOpenTelemetry(
            logging => logging.AddOtlpExporter(o => o.Endpoint = new Uri("http://myotlpendpoint:4317/logs")),
            options => options.IncludeFormattedMessage = true));
    ```
    I think we should not include it with this change. As per my understanding, it depends on the final outcome of https://github.com/open-telemetry/opentelemetry-dotnet/pull/5325. If we go ahead with the OpenTelemetrySdk pattern then we could just ask users to follow that (Separate discussion).

In short, the only use case we would enable right now is `.WithLogging()` for hosting scenarios.

For future, we could decide whether we want to have the migration guidance for case 2 or go with OpenTelemetrySdk pattern.

**Comment by CodeBlanch** [comment]
@vishweshbankwar 

> @CodeBlanch - I think we should not mark things obsolete rn.

Based on what thinking/reasons? Please elaborate a bit 😄

**Comment by vishweshbankwar** [comment]
> @vishweshbankwar
> 
> > @CodeBlanch - I think we should not mark things obsolete rn.
> 
> Based on what thinking/reasons? Please elaborate a bit 😄

* Based on my point #2: Once we have a clear plan/alternate for users on case 2 then we can consider the obsoleting part. 
* Also, following this staged approach would give enough time for library authors also to support the extensions based on `LoggerProviderBuilder`.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs [hunk 1]
```diff
@@ -28,6 +25,7 @@ public static class OtlpLogExporterHelperExtensions
     /// <remarks><inheritdoc cref="AddOtlpExporter(OpenTelemetryLoggerOptions, Action{OtlpExporterOptions})" path="/remarks"/></remarks>
     /// <param name="loggerOptions"><see cref="OpenTelemetryLoggerOptions"/> options to use.</param>
     /// <returns>The instance of <see cref="OpenTelemetryLoggerOptions"/> to chain the calls.</returns>
+    [Obsolete("Call LoggerProviderBuilder.AddOtlpExporter instead this method will be removed in a future version.")]
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:28]: 👍 to obsoleting these methods now. Obsoleting is the signal used to push users to new, preferred APIs. Without the signal, it remains unclear which API to use. Obsoleting is better than alternatives like relying on people to read release notes or intellisense documentation - they often don't.

File: src/OpenTelemetry/.publicApi/Experimental/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -24,20 +22,5 @@ OpenTelemetry.Metrics.MetricStreamConfiguration.CardinalityLimit.set -> void
...
-static OpenTelemetry.Logs.LoggerProviderExtensions.ForceFlush(this OpenTelemetry.Logs.LoggerProvider! provider, int timeoutMilliseconds = -1) -> bool
-static OpenTelemetry.Logs.LoggerProviderExtensions.Shutdown(this OpenTelemetry.Logs.LoggerProvider! provider, int timeoutMilliseconds = -1) -> bool
-static OpenTelemetry.OpenTelemetryBuilderSdkExtensions.WithLogging(this OpenTelemetry.IOpenTelemetryBuilder! builder) -> OpenTelemetry.IOpenTelemetryBuilder!
-static OpenTelemetry.OpenTelemetryBuilderSdkExtensions.WithLogging(this OpenTelemetry.IOpenTelemetryBuilder! builder, System.Action<OpenTelemetry.Logs.LoggerProviderBuilder!>! configure) -> OpenTelemetry.IOpenTelemetryBuilder!
-static OpenTelemetry.OpenTelemetryBuilderSdkExtensions.WithLogging(this OpenTelemetry.IOpenTelemetryBuilder! builder, System.Action<OpenTelemetry.Logs.LoggerProviderBuilder!>? configureBuilder, System.Action<OpenTelemetry.Logs.OpenTelemetryLoggerOptions!>? configureOptions) -> OpenTelemetry.IOpenTelemetryBuilder!
 static OpenTelemetry.Sdk.CreateLoggerProviderBuilder() -> OpenTelemetry.Logs.LoggerProviderBuilder!
```
Comment (alanwest) [inline src/OpenTelemetry/.publicApi/Experimental/PublicAPI.Unshipped.txt:25]: I'm ok keeping this experimental for now. Might ultimately be nice simply for consistency, but I see no reason for this discussion to block the introduction of the new APIs, especially `WithLogging`.

**Comment by CodeBlanch** [comment]
We decided on the SIG meeting 6/4/2024 that we would:

* Not obsolete anything for `1.9.0`. We will obsolete things in a future release. This was done to give library authors time to add extensions targeting `LoggerProviderBuilder` which will ease migration for users when things are obsoleted.

* `ILoggingBuilder.UseOpenTelemetry` will not be exposed initially. It may be replaced by the design on #5325.

@alanwest @vishweshbankwar PR has been updated.

File: src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+abstract OpenTelemetry.Logs.LoggerProviderBuilder.AddInstrumentation<TInstrumentation>(System.Func<TInstrumentation>! instrumentationFactory) -> OpenTelemetry.Logs.LoggerProviderBuilder!
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt:1]: Just curious, When would `AddInstrumentation` be needed with respect to logging?

Comment (cijothomas) [reply src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt:1]: Whenever the instrumentation libraries start producing logs. As of today, only traces and metrics are emitted by most, but there is nothing preventing an instrumentation library to emit logs too. Of course, it requires an `ILoggerFactory` or an `ILogger`  to be given to the instrumentation library (unlike Meter,ActivitySource, which can be created out of thin air using ctor() )

Comment (vishweshbankwar) [reply src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt:1]: Got it - I was curious to know if there are any specific examples where this is needed today.

File: src/OpenTelemetry.Api/Logs/LoggerProvider.cs [hunk 4]
```diff
@@ -6,28 +6,16 @@
...
-#endif
-public
-#else
-/// <summary>
-/// LoggerProvider is the entry point of the OpenTelemetry API. It provides access to <see cref="Logger"/>.
-/// </summary>
-internal
-#endif
-    class LoggerProvider : BaseProvider
+public class LoggerProvider : BaseProvider
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Api/Logs/LoggerProvider.cs:18]: Is this needed outside of hosting package?

File: src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt:1]: no specific example I am aware of. Most libraries are natively producing Ilogger logs today, so there wasn't a pressing demand for an instrumentation library for logs.

**Comment by cijothomas** [comment]
@CodeBlanch Can you update the PR description to include the changes expected to be done by end users as well, even though they are not *forced* to do it? Also call out that, this should be done only if their exporter supports this new model (and mention that exporters shipped from this repo already supports this model)

File: examples/AspNetCore/Program.cs [hunk 5]
```diff
@@ -121,34 +122,25 @@
...
-appBuilder.Logging.ClearProviders();
-
-// Configure OpenTelemetry Logging.
-appBuilder.Logging.AddOpenTelemetry(options =>
-{
-    // Note: See appsettings.json Logging:OpenTelemetry section for configuration.
-
-    var resourceBuilder = ResourceBuilder.CreateDefault();
-    configureResource(resourceBuilder);
-    options.SetResourceBuilder(resourceBuilder);
```
Comment (cijothomas) [inline examples/AspNetCore/Program.cs:136]: ❤️ seeing common Resource across all 3!

And also ❤️ fixing the awkwardness of calling methods like `AddExporter` on a supposedly pure "Options" class!

File: docs/logs/getting-started-aspnetcore/Program.cs [hunk 6]
```diff
@@ -6,25 +6,21 @@
...
-        .CreateDefault()
-        .AddService(builder.Environment.ApplicationName);
-
-    logging.SetResourceBuilder(resourceBuilder)
-
-        // ConsoleExporter is used for demo purpose only.
-        // In production environment, ConsoleExporter should be replaced with other exporters (e.g. OTLP Exporter).
-        .AddConsoleExporter();
-});
+// Add OpenTelemetry logging provider by calling the WithLogging extension.
```
Comment (cijothomas) [inline docs/logs/getting-started-aspnetcore/Program.cs:16]: should we add a note here to warn users that this approach only works if their exporter also supports this model?

File: src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt:1]: The other nice thing `AddInstrumentation` does is allow you to attach things to the lifetime of a provider. When disposed the provider will in turn call `Dispose` on anything added as instrumentation (if it implements `IDisposable`).

I used that way back in the [EventSource -> Logs PoC](https://github.com/open-telemetry/opentelemetry-dotnet/blob/5be8226861c12b021bc1ea65516ea05755e8c1f9/src/OpenTelemetry.Extensions.EventSource/OpenTelemetryEventSourceLoggerProviderBuilderExtensions.cs#L78-L83).

Probably won't be used a whole lot but good to keep logs/metrics/tracing providers consistent in that way?

File: src/OpenTelemetry.Api/Logs/LoggerProvider.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Logs/LoggerProvider.cs:18]: The main thing exposing this does is let users call `ForceFlush` if they need to. They can also call `Shutdown` or `AddProcessor` if needed. None of those scenarios are possible today so it is mostly a spec compliance thing.

How would you do that? Something like this...

```csharp
builder.Services.AddOpenTelemetry()
    .WithLogging(logging => logging.AddOtlpExporter());

var app = builder.Build();

var loggerProvider = app.Services.GetRequiredService<LoggerProvider>();

loggerProvider.ForceFlush();
```

PS: The Log Bridge API is all rooted off `LoggerProvider` too but that isn't exposed yet.

**Comment by CodeBlanch** [comment]
@cijothomas I added "Migration guidance for users" back on the PR description.

File: src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt (hunk 3 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Unshipped.txt:1]: Sounds good to me 👍

File: docs/logs/getting-started-aspnetcore/Program.cs (hunk 6 above)
Comment (CodeBlanch) [reply docs/logs/getting-started-aspnetcore/Program.cs:16]: Might get confusing. I think either a) we leave it as is or b) we show the old style. Let me know which you prefer! I'm hoping the gap is short-lived and we can get these extensions added soon after 1.9.0 is out.

**Review by vishweshbankwar — APPROVED** [review]
LGTM

File: docs/logs/getting-started-aspnetcore/Program.cs (hunk 6 above)
Comment (cijothomas) [reply docs/logs/getting-started-aspnetcore/Program.cs:16]: okay to keep this as-is, assuming the gap is short-lived.
